principal.unwrap is not a function

### Environment info
- NooBaa Version: nsfs / noobaa-core-5.15.0-1.el8.x86_64 (head commit 08c1d008ab482830bf9b381cf073a469e409effb)
- Platform: RHEL 8.6, GPFS 5.1.7.1, NSFS

### Actual behavior
Bucket policy doesn't work. Getting `TypeError: principal.unwrap is not a function` when trying to access.

### Steps to reproduce
Create policy file with content:

```
{
"Version":"2012-10-17",
"Statement":[
        {
          "Sid":"this-is-the-sid",
          "Effect":"Allow",
          "Principal": [
              "actor2@noobaa.io"
           ],
          "Action":["s3:*"],
          "Resource":["arn:aws:s3:::*"]
        }
   ]
}
```

(have tried various ways of pointing at the principal.. name, email, "*".. always same error)

Assign to bucket:

`aws --profile actor3 --endpoint http://localhost:6001  s3api put-bucket-policy --bucket ns9012k --policy  file:///root/allow-actor2-ns9012k.policy`

Access bucket as actor2:

```
# aws --profile actor2 --endpoint http://localhost:6001 s3 ls s3://ns9012k`

An error occurred (InternalError) when calling the ListObjectsV2 operation (reached max retries: 2): We encountered an internal error. Please try again.`
```

### More information - Screenshots / Logs / Other output

Strange it's saying validate_non_nsfs_bucket in the log, as this is supposed to be NSFS.

```
Oct 06 14:42:41 scale-test-os-1.oslo.forum.ibm.com node[3449126]: Oct-6 14:42:41.511 [nsfs/3449126]    [L1] core.sdk.object_sdk:: validate_non_nsfs_bucket:  { name: SENSITIVE-f4532c518b6d03dc, email: SENSITIVE-f4532c518b6d03dc, has_login: 'false', has_s3_access: 'true', allow_bucket_creation: 'true', access_keys: [ { access_key: SENSITIVE-92c409ac6a89ab37, secret_key: SENSITIVE-0683bcd1285ff91f } ], nsfs_account_config: { uid: 5678, gid: 5678, new_buckets_path: '/ibm/fs1/buckets1', nsfs_only: 'true' } } { fs_root_path: '' }
Oct 06 14:42:41 scale-test-os-1.oslo.forum.ibm.com node[3449126]: Oct-6 14:42:41.513 [nsfs/3449126] [ERROR] core.endpoint.s3.s3_rest:: S3 ERROR <?xml version="1.0" encoding="UTF-8"?><Error><Code>InternalError</Code><Message>We encountered an internal error. Please try again.</Message><Resource>/ns9012k?list-type=2&amp;prefix=&amp;delimiter=%2F&amp;encoding-type=url</Resource><RequestId>lneljsgz-elrq7f-19z5</RequestId></Error> GET /ns9012k?list-type=2&prefix=&delimiter=%2F&encoding-type=url {"host":"localhost:6001","accept-encoding":"identity","user-agent":"aws-cli/2.4.6 Python/3.8.8 Linux/4.18.0-348.2.1.el8_5.x86_64 exe/x86_64.rhel.8 prompt/off command/s3.ls","x-amz-date":"20231006T124241Z","x-amz-content-sha256":"e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855","authorization":"AWS4-HMAC-SHA256 Credential=kWmbUkx99didgIw27HqG/20231006/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=a05a4fb8584191d9690028cb4380ab923d4b8a4ba35e7456121373bfc16f9f98"} TypeError: principal.unwrap is not a function
Oct 06 14:42:41 scale-test-os-1.oslo.forum.ibm.com node[3449126]:     at _is_principal_fit (/usr/local/noobaa-core/src/endpoint/s3/s3_bucket_policy_utils.js:168:24)
Oct 06 14:42:41 scale-test-os-1.oslo.forum.ibm.com node[3449126]:     at _is_statements_fit (/usr/local/noobaa-core/src/endpoint/s3/s3_bucket_policy_utils.js:192:31)
Oct 06 14:42:41 scale-test-os-1.oslo.forum.ibm.com node[3449126]:     at Object.has_bucket_policy_permission (/usr/local/noobaa-core/src/endpoint/s3/s3_bucket_policy_utils.js:143:15)
Oct 06 14:42:41 scale-test-os-1.oslo.forum.ibm.com node[3449126]:     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
Oct 06 14:42:41 scale-test-os-1.oslo.forum.ibm.com node[3449126]:     at async authorize_request_policy (/usr/local/noobaa-core/src/endpoint/s3/s3_rest.js:237:24)
Oct 06 14:42:41 scale-test-os-1.oslo.forum.ibm.com node[3449126]:     at async Promise.all (index 1)
Oct 06 14:42:41 scale-test-os-1.oslo.forum.ibm.com node[3449126]:     at async authorize_request (/usr/local/noobaa-core/src/endpoint/s3/s3_rest.js:197:5)
Oct 06 14:42:41 scale-test-os-1.oslo.forum.ibm.com node[3449126]:     at async handle_request (/usr/local/noobaa-core/src/endpoint/s3/s3_rest.js:113:5)
Oct 06 14:42:41 scale-test-os-1.oslo.forum.ibm.com node[3449126]:     at async Object.s3_rest [as handler] (/usr/local/noobaa-core/src/endpoint/s3/s3_rest.js:64:9)
```

--- a/src/endpoint/s3/s3_bucket_policy_utils.js
+++ b/src/endpoint/s3/s3_bucket_policy_utils.js
@@ -4,6 +4,7 @@
 const _ = require('lodash');
 const dbg = require('../../util/debug_module')(__filename);
 const s3_utils = require('./s3_utils');
+const RpcError = require('../../rpc/rpc_error');
 
 const OP_NAME_TO_ACTION = Object.freeze({
     delete_bucket_analytics: { regular: "s3:PutAnalyticsConfiguration" },
@@ -115,6 +116,8 @@ const supported_actions = {
     's3:ExistingObjectTag': ['s3:DeleteObjectTagging', 's3:DeleteObjectVersionTagging', 's3:GetObject', 's3:GetObjectAcl', 's3:GetObjectTagging', 's3:GetObjectVersion', 's3:GetObjectVersionTagging', 's3:PutObjectAcl', 's3:PutObjectTagging', 's3:PutObjectVersionTagging'],
     's3:x-amz-server-side-encryption': ['s3:PutObject']
 };
+
+const SUPPORTED_BUCKET_POLICY_CONDITIONS = Object.keys(supported_actions);
 
 async function _is_server_side_encryption_fit(req, predicate, value) {
     const encryption = s3_utils.parse_encryption(req);
@@ -233,7 +236,51 @@ function _parse_condition_keys(condition_statement) {
     }
 }
 
+function validate_s3_policy(policy, bucket_name, get_account_handler) {
+    const all_op_names = _.compact(_.flatMap(OP_NAME_TO_ACTION, action => [action.regular, action.versioned]));
+    for (const statement of policy.Statement) {
+
+        const statement_principal = statement.Principal || statement.NotPrincipal;
+        if (statement_principal.AWS) {
+            for (const principal of _.flatten([statement_principal.AWS])) {
+                if ((typeof principal === 'string') ? principal !== '*' : principal.unwrap() !== '*') {
+                    const account = get_account_handler(principal);
+                    if (!account) {
+                        throw new RpcError('MALFORMED_POLICY', 'Invalid principal in policy', { detail: principal });
+                    }
+                }
+            }
+        } else if ((typeof statement_principal === 'string') ? statement_principal !== '*' : statement_principal.unwrap() !== '*') {
+            throw new RpcError('MALFORMED_POLICY', 'Invalid principal in policy', { detail: statement.Principal });
+        }
+        for (const resource of _.flatten([statement.Resource || statement.NotResource])) {
+            const resource_bucket_part = resource.split('/')[0];
+            const resource_regex = RegExp(`^${resource_bucket_part.replace(qm_regex, '.?').replace(ar_regex, '.*')}$`);
+            if (!resource_regex.test('arn:aws:s3:::' + bucket_name)) {
+                throw new RpcError('MALFORMED_POLICY', 'Policy has invalid resource', { detail: resource });
+            }
+        }
+        for (const action of _.flatten([statement.Action || statement.NotAction])) {
+            if (action !== 's3:*' && !all_op_names.includes(action)) {
+                throw new RpcError('MALFORMED_POLICY', 'Policy has invalid action', { detail: action });
+            }
+        }
+        if (statement.Condition) {
+            for (const condition of Object.values(statement.Condition)) {
+                for (const condition_key of Object.keys(condition)) {
+                    // some condition keys have arguments in their names.(e.g. s3:ExistingObjectTag/<key>)
+                    // parse to get only the condition key itself
+                    const key_parts = condition_key.split("/");
+                    if (!SUPPORTED_BUCKET_POLICY_CONDITIONS.includes(key_parts[0])) {
+                        throw new RpcError('MALFORMED_POLICY', 'Policy has invalid condition key or unsupported condition key', { detail: condition_key });
+                    }
+                }
+            }
+        }
+        // TODO: Need to validate that the resource comply with the action
+    }
+}
 
 exports.OP_NAME_TO_ACTION = OP_NAME_TO_ACTION;
-exports.SUPPORTED_BUCKET_POLICY_CONDITIONS = Object.keys(supported_actions);
 exports.has_bucket_policy_permission = has_bucket_policy_permission;
+exports.validate_s3_policy = validate_s3_policy;

--- a/src/sdk/bucketspace_fs.js
+++ b/src/sdk/bucketspace_fs.js
@@ -13,6 +13,7 @@ const P = require('../util/promise');
 const BucketSpaceSimpleFS = require('./bucketspace_simple_fs');
 const _ = require('lodash');
 const util = require('util');
+const bucket_policy_utils = require('../endpoint/s3/s3_bucket_policy_utils');
 
 const dbg = require('../util/debug_module')(__filename);
 
@@ -129,6 +130,18 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
 
             bucket.system_owner = new SensitiveString(bucket.system_owner);
             bucket.bucket_owner = new SensitiveString(bucket.bucket_owner);
+            if (bucket.s3_policy) {
+                for (const [s_index, statement] of bucket.s3_policy.Statement.entries()) {
+                    const statement_principal = statement.Principal || statement.NotPrincipal;
+                    if (statement_principal.AWS) {
+                        const sensitive_arr = _.flatten([statement_principal.AWS]).map(principal => new SensitiveString(principal));
+                        if (statement.Principal) bucket.s3_policy.Statement[s_index].Principal.AWS = sensitive_arr;
+                        if (statement.NotPrincipal) bucket.s3_policy.Statement[s_index].NotPrincipal.AWS = sensitive_arr;
+                    } else {
+                        bucket.s3_policy.Statement = new SensitiveString(statement_principal);
+                    }
+                }
+            }
             return bucket;
         } catch (err) {
             throw this._translate_object_error_codes(err);
@@ -512,6 +525,7 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
             const bucket_config_path = this._get_bucket_config_path(name);
             const { data } = await nb_native().fs.readFile(this.fs_context, bucket_config_path);
             const bucket = JSON.parse(data.toString());
+            bucket_policy_utils.validate_s3_policy(policy, bucket.name, () => true);
             bucket.s3_policy = policy;
             const update_bucket = JSON.stringify(bucket);
             await nb_native().fs.writeFile(

--- a/src/server/system_services/bucket_server.js
+++ b/src/server/system_services/bucket_server.js
@@ -33,7 +33,7 @@ const azure_storage = require('../../util/azure_storage_wrap');
 const usage_aggregator = require('../bg_services/usage_aggregator');
 const chunk_config_utils = require('../utils/chunk_config_utils');
 const NetStorage = require('../../util/NetStorageKit-Node-master/lib/netstorage');
-const { OP_NAME_TO_ACTION, SUPPORTED_BUCKET_POLICY_CONDITIONS } = require('../../endpoint/s3/s3_bucket_policy_utils');
+const bucket_policy_utils = require('../../endpoint/s3/s3_bucket_policy_utils');
 const path = require('path');
 const KeysSemaphore = require('../../util/keys_semaphore');
 const bucket_semaphore = new KeysSemaphore(1);
@@ -47,8 +47,6 @@ const VALID_BUCKET_NAME_REGEXP =
 const EXTERNAL_BUCKET_LIST_TO = 30 * 1000; //30s
 
 const trigger_properties = ['event_name', 'object_prefix', 'object_suffix'];
-const qm_regex = /\?/g;
-const ar_regex = /\*/g;
 
 function new_bucket_defaults(name, system_id, tiering_policy_id, owner_account_id, tag, lock_enabled) {
     const now = Date.now();
@@ -487,7 +485,7 @@ async function get_bucket_policy(req) {
 async function put_bucket_policy(req) {
     dbg.log0('put_bucket_policy:', req.rpc_params);
     const bucket = find_bucket(req, req.rpc_params.name);
-    _validate_s3_policy(req.rpc_params.policy, bucket.name);
+    bucket_policy_utils.validate_s3_policy(req.rpc_params.policy, bucket.name, principal => system_store.get_account_by_email(principal));
     await system_store.make_changes({
         update: {
             buckets: [{
@@ -498,50 +496,6 @@ async function put_bucket_policy(req) {
     });
 }
 
-function _validate_s3_policy(policy, bucket_name) {
-    const all_op_names = _.compact(_.flatMap(OP_NAME_TO_ACTION, action => [action.regular, action.versioned]));
-    for (const statement of policy.Statement) {
-
-        const statement_principal = statement.Principal || statement.NotPrincipal;
-        if (statement_principal.AWS) {
-            for (const principal of _.flatten([statement_principal.AWS])) {
-                if (principal.unwrap() !== '*') {
-                    const account = system_store.get_account_by_email(principal);
-                    if (!account) {
-                        throw new RpcError('MALFORMED_POLICY', 'Invalid principal in policy', { detail: principal });
-                    }
-                }
-            }
-        } else if (statement_principal.unwrap() !== '*') {
-            throw new RpcError('MALFORMED_POLICY', 'Invalid principal in policy', { detail: statement.Principal });
-        }
-        for (const resource of _.flatten([statement.Resource || statement.NotResource])) {
-            const resource_bucket_part = resource.split('/')[0];
-            const resource_regex = RegExp(`^${resource_bucket_part.replace(qm_regex, '.?').replace(ar_regex, '.*')}$`);
-            if (!resource_regex.test('arn:aws:s3:::' + bucket_name)) {
-                throw new RpcError('MALFORMED_POLICY', 'Policy has invalid resource', { detail: resource });
-            }
-        }
-        for (const action of _.flatten([statement.Action || statement.NotAction])) {
-            if (action !== 's3:*' && !all_op_names.includes(action)) {
-                throw new RpcError('MALFORMED_POLICY', 'Policy has invalid action', { detail: action });
-            }
-        }
-        if (statement.Condition) {
-            for (const condition of Object.values(statement.Condition)) {
-                for (const condition_key of Object.keys(condition)) {
-                    // some condition keys have arguments in their names.(e.g. s3:ExistingObjectTag/<key>)
-                    // parse to get only the condition key itself
-                    const key_parts = condition_key.split("/");
-                    if (!SUPPORTED_BUCKET_POLICY_CONDITIONS.includes(key_parts[0])) {
-                        throw new RpcError('MALFORMED_POLICY', 'Policy has invalid condition key or unsupported condition key', { detail: condition_key });
-                    }
-                }
-            }
-        }
-        // TODO: Need to validate that the resource comply with the action
-    }
-}
 
 
 async function delete_bucket_policy(req) {

--- a/src/test/unit_tests/test_bucketspace_fs.js
+++ b/src/test/unit_tests/test_bucketspace_fs.js
@@ -309,6 +309,48 @@ mocha.describe('bucketspace_fs', function() {
             const output_web = await bucketspace_fs.get_bucket_policy(param);
             assert.ok(output_web.policy === undefined);
         });
+
+        mocha.it('put_bucket_policy other account object', async function() {
+            const policy = {
+                    Version: '2012-10-17',
+                    Statement: [{
+                        Sid: 'id-22',
+                        Effect: 'Allow',
+                        Principal: { AWS: 'noobaa@noobaa.io' },
+                        Action: ['s3:*'],
+                        Resource: ['arn:aws:s3:::*']
+                        }
+                    ]
+                };
+            const param = {name: test_bucket, policy: policy};
+            await bucketspace_fs.put_bucket_policy(param);
+            const bucket_policy = await bucketspace_fs.get_bucket_policy(param);
+            assert.deepEqual(bucket_policy.policy, policy);
+        });
+
+        mocha.it('put_bucket_policy other account array', async function() {
+            const policy = {
+                    Version: '2012-10-17',
+                    Statement: [{
+                        Sid: 'id-22',
+                        Effect: 'Allow',
+                        Principal: { AWS: ['noobaa@noobaa.io', 'noobaa1@noobaa.io'] },
+                        Action: ['s3:*'],
+                        Resource: ['arn:aws:s3:::*']
+                        }
+                    ]
+                };
+            const param = {name: test_bucket, policy: policy};
+            await bucketspace_fs.put_bucket_policy(param);
+            const bucket_policy = await bucketspace_fs.get_bucket_policy(param);
+            assert.deepEqual(bucket_policy.policy, policy);
+        });
+        mocha.it('delete_bucket_policy ', async function() {
+            const param = {name: test_bucket};
+            await bucketspace_fs.delete_bucket_policy(param);
+            const bucket_policy = await bucketspace_fs.get_bucket_policy(param);
+            assert.ok(bucket_policy.policy === undefined);
+        });
     });
 });
 

